src/lib/sql/sql_pipeline.cpp

```diff
@@ -217,22 +225,38 @@ std::pair<SQLPipelineStatus, const std::vector<std::shared_ptr<const Table>>&> S
   _result_tables.reserve(_sql_pipeline_statements.size());
 
   for (auto& pipeline_statement : _sql_pipeline_statements) {
+    pipeline_statement->set_transaction_context(_transaction_context);
     const auto& [statement_status, table] = pipeline_statement->get_result_table();
-    if (statement_status == SQLPipelineStatus::RolledBack) {
+    if (statement_status == SQLPipelineStatus::Failure) {
       _failed_pipeline_statement = pipeline_statement;
 
-      if (_transaction_context) {
+      if (_transaction_context && !_transaction_context->is_auto_commit()) {
         // The pipeline was executed using a transaction context (i.e., no auto-commit after each statement).
         // Previously returned results are invalid.
         _result_tables.clear();
+        _transaction_context = nullptr;
       }
 
-      return {SQLPipelineStatus::RolledBack, _result_tables};
+      return {SQLPipelineStatus::Failure, _result_tables};
     }
 
     DebugAssert(statement_status == SQLPipelineStatus::Success, "Unexpected pipeline status");
 
     _result_tables.emplace_back(table);
+
+    const auto& previous_transaction_context = pipeline_statement->transaction_context();
+
+    if (previous_transaction_context->is_auto_commit()) {
+      DebugAssert(previous_transaction_context->phase() == TransactionPhase::Committed ||
+                      previous_transaction_context->phase() == TransactionPhase::RolledBackByUser,
```

Comment:
Can an auto commit transaction be in TransactionPhase::RolledBackByUser? If not, should _transition guard against this?